tensorflow/c/eager/dlpack.cc

```diff
@@ -0,0 +1,292 @@
+  TFDLManagedTensorCtx* owner =
+      static_cast<TFDLManagedTensorCtx*>(arg->manager_ctx);
+  owner->handle->Unref();
+  delete owner;
+}
+
+DLDataType GetDLDataType(TF_DataType data_type, TF_Status* status) {
+  DLDataType dtype;
+  dtype.lanes = 1;
+  dtype.bits = TF_DataTypeSize(data_type) * 8;
+  switch (data_type) {
+    case TF_DataType::TF_HALF:
+    case TF_DataType::TF_FLOAT:
+    case TF_DataType::TF_DOUBLE:
+      dtype.code = DLDataTypeCode::kDLFloat;
+      break;
+    case TF_DataType::TF_INT8:
+    case TF_DataType::TF_INT16:
+    case TF_DataType::TF_INT32:
+    case TF_DataType::TF_INT64:
+      dtype.code = DLDataTypeCode::kDLInt;
+      break;
+    case TF_DataType::TF_BOOL:
+    case TF_DataType::TF_UINT8:
+    case TF_DataType::TF_UINT16:
+    case TF_DataType::TF_UINT32:
+    case TF_DataType::TF_UINT64:
+      dtype.code = DLDataTypeCode::kDLUInt;
+      break;
+    case TF_DataType::TF_BFLOAT16:
+      dtype.code = DLDataTypeCode::kDLBfloat;
+      break;
+    default:
+      status->status = tensorflow::errors::InvalidArgument(
+          DataType_Name(static_cast<DataType>(data_type)),
+          " is not supported by dlpack");
+      break;
+  }
+  return dtype;
+}
+
+DLContext GetDLContext(TFE_TensorHandle* h, TF_Status* status) {
+  DLContext ctx;
+  const char* device_name = h->handle->DeviceName(&status->status);
+  DeviceNameUtils::ParsedName parsed_name;
+  tensorflow::DeviceNameUtils::ParseFullName(device_name, &parsed_name);
+  std::string device_type = parsed_name.type;
+  int device_id = -1;
+  if (parsed_name.has_id) {
+    device_id = parsed_name.id;
+  }  // Question: Is it possible that it doens't have id?
+
+  ctx.device_id = device_id;
+  if (device_type == "CPU") {
+    ctx.device_type = DLDeviceType::kDLCPU;
+  } else if (device_type == "GPU") {
+    ctx.device_type = DLDeviceType::kDLGPU;
+  } else {
+    status->status = tensorflow::errors::InvalidArgument(
+        "Unsupported Device Type for dlpack");
+  }
+
+  return ctx;
+}
+
+DLManagedTensor* TFEHandleToTFDLManagedTensorCtx(TFE_TensorHandle* h,
+                                                 TF_Status* status) {
+  const Tensor* tensor = GetTensorFromHandle(h, status);
+  TF_DataType data_type = static_cast<TF_DataType>(tensor->dtype());
+  auto* tf_dlm_tensor_ctx = new TFDLManagedTensorCtx;
+
+  TensorReference* tensor_ref =
+      new TensorReference(*tensor);  // This will call buf_->Ref()
+  tf_dlm_tensor_ctx->handle = tensor_ref;
+  tf_dlm_tensor_ctx->tensor.manager_ctx = tf_dlm_tensor_ctx;
+  tf_dlm_tensor_ctx->tensor.deleter = &DLManagedTensorDeleter;
+  tf_dlm_tensor_ctx->tensor.dl_tensor.ctx = GetDLContext(h, status);
+  int ndim = tensor->dims();
+  tf_dlm_tensor_ctx->tensor.dl_tensor.ndim = ndim;
+  tf_dlm_tensor_ctx->tensor.dl_tensor.data =
+      TFE_TensorHandleDevicePointer(h, status);
+  tf_dlm_tensor_ctx->tensor.dl_tensor.dtype = GetDLDataType(data_type, status);
+
+  int64_t* shape_arr = new int64_t[ndim];
```

Comment:
This appears to be leaked? I don't see any logic to free the shape.

Reply:
Thanks. You are right. I'll fix this.

Reply:
I've fixed this by using vector as ctx members. (Refer to your code for jax)